### PR TITLE
Fix RUL DistSQL execution failure in transaction(#19831)

### LINKED ISSUE #19831: Unable to execute preview command normally in transaction
## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?
master

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
proxy

### Expected behavior
preview can be executed successfully

### Actual behavior
U0ZC0EFquu

### Reason analyze (If you can)
```
java.lang.UnsupportedOperationException: Non-query dist sql is not supported within a transaction
	at org.apache.shardingsphere.proxy.backend.handler.ProxyBackendHandlerFactory.newInstance(ProxyBackendHandlerFactory.java:110)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.<init>(MySQLComQueryPacketExecutor.java:72)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.MySQLCommandExecutorFactory.newInstance(MySQLCommandExecutorFactory.java:77)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.MySQLCommandExecuteEngine.getCommandExecutor(MySQLCommandExecuteEngine.java:61)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:110)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:78)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
```sql
begin;
preview xxx;
```

### PR DESCRIPTION
**Pull request by FlyingZC**
Fixes #19831.

Changes proposed in this pull request:
- Fix RUL DistSQL execution failure in transaction

### CODE REVIEW

**Comment by FlyingZC** [comment]
Execution result after bug fix:
```sql
mysql> use readwrite_splitting_db;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
mysql> begin;
Query OK, 0 rows affected (0.03 sec)

mysql> preview insert into account values(1,1,1);
+------------------+-----------------------------------+
| data_source_name | actual_sql                        |
+------------------+-----------------------------------+
| write_ds         | insert into account values(1,1,1) |
+------------------+-----------------------------------+
1 row in set (0.09 sec)
```

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java [hunk 1]
```diff
@@ -106,16 +107,14 @@ public static ProxyBackendHandler newInstance(final DatabaseType databaseType, f
...
-            }
+            checkUnsupportedDistSQLStatementInTransaction(sqlStatement, connectionSession);
             return DistSQLBackendHandlerFactory.newInstance((DistSQLStatement) sqlStatement, connectionSession);
         }
         SQLStatementContext<?> sqlStatementContext = SQLStatementContextFactory.newInstance(ProxyContext.getInstance().getContextManager().getMetaDataContexts().getMetaData().getDatabases(),
                 sqlStatement, connectionSession.getDefaultDatabaseName());
         return newInstance(databaseType, new LogicSQL(sqlStatementContext, sql, Collections.emptyList()), connectionSession, false);
     }
-    
+
```
Comment (RaigorJiang) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java:117]: Please keep the indent consistent with the previous line.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java [hunk 2]
```diff
@@ -160,6 +159,16 @@ public static ProxyBackendHandler newInstance(final DatabaseType databaseType, f
...
         return backendHandler.orElseGet(() -> DatabaseBackendHandlerFactory.newInstance(logicSQL, connectionSession, preferPreparedStatement));
     }
+
+    private static void checkUnsupportedDistSQLStatementInTransaction(final SQLStatement sqlStatement, final ConnectionSession connectionSession) {
+        if (connectionSession.getTransactionStatus().isInTransaction() && !isSupportedDistSQLStatement(sqlStatement)) {
+            throw new UnsupportedOperationException("Non-query dist sql is not supported within a transaction");
+        }
+    }
+
+    private static boolean isSupportedDistSQLStatement(final SQLStatement sqlStatement) {
```
Comment (RaigorJiang) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java:169]: Could `isSupportedDistSQLStatementInTransaction` be better?

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java (hunk 1 above)
Comment (FlyingZC) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java:117]: ok

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java (hunk 2 above)
Comment (FlyingZC) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java:169]: Your are right.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java (hunk 1 above)
Comment (FlyingZC) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/handler/ProxyBackendHandlerFactory.java:117]: image

The content of this line is the same as before, the modification of this line you see comes from my first commit.

**Comment by RaigorJiang** [comment]
@FlyingZC 
You may not be using the IDEA configuration provided by the community, which is prone to some formatting issues.

image

**Comment by RaigorJiang** [comment]
Please refer to https://shardingsphere.apache.org/community/en/contribute/code-conduct/

image

**Comment by FlyingZC** [comment]
> Please refer to https://shardingsphere.apache.org/community/en/contribute/code-conduct/
> 
> image

I have formatted the code by spotless fromat, the settings maybe have some problems.

**Review by RaigorJiang — APPROVED** [review]

**Pull request merged**